README.md
# FSX simconnect-as-a-service

FSX informations exposed by a JSON HTTP API server

This can be used to interface other programs with a running FSX instance via HTTP requests.

The exposed data is read-only (e.g. only LIST and RETRIEVE actions are supported).

An example of use of this application is [FSX web ATC](https://marcosox.github.io/fsx-web-atc/), which shows the data in a web map.

## Download and installation
FSX-saas is portable, you just need to download the latest jar from the [releases page] and run it.

## Usage
    java -jar path/to/fsx-saas-x.y.z.jar
    
To change the application parameters pass the path of a json configuration file:

    java -jar path/to/fsx-saas-x.y.z.jar -conf path/to/config.json

The complete configuration file is this:

    {
        "port": 8080
        "scanInterval": 1000
    }

- `port`: tcp listening port for the server
- `scanInterval`: milliseconds between aircraft requests to fsx

#### Vertx options
Since this application is packaged with a Vertx launcher, all the vertx options can be passed.
For more informations see the [help page](http://vertx.io/docs/vertx-core/java/#_the_vertx_command_line)

## Quick start and examples
- Start FSX and load a flight
- Start FSX-saas
- open your browser and point it to [http://localhost:8080/](http://localhost:8080/) to see all the available API endpoints
- See all the aircrafts: [http://localhost:8080/aircrafts](http://localhost:8080/aircrafts)
```
[ {
  "id" : 990,
  "title" : "de Havilland Dash 8-100 Paint2",
  "atcType" : "DeHavilland",
  "atcModel" : "DH8A",
  "atcID" : "N7048R",
  "atcAirline" : "American Pacific",
  "atcFlightNumber" : "1123",
  "from" : "KBFI",
  "to" : "KIDA",
  "latitude" : 47.524769343755345,
  "longitude" : -122.29564771099874,
  "altitude" : 25.338651001296956,
  "altAgl" : 7.339984247004242,
  "airspeed" : 0,
  "pitch" : 0.002319927327334881,
  "bank" : 0.0,
  "heading" : 56.78009212492774,
  "aileron" : 0.0,
  "elevator" : 0.0,
  "rudder" : 0.0,
  "throttle" : 0.0
}, {
...
```
- See user aircraft: [http://localhost:8080/aircrafts/0](http://localhost:8080/aircrafts/0)
```
{
  "id" : 0,
  "title" : "Aircreation582SL red",
  "atcType" : "Ultralight",
  "atcModel" : "trike",
  "atcID" : "",
  "atcAirline" : "",
  "atcFlightNumber" : "",
  "from" : "",
  "to" : "",
  "latitude" : 48.502472910349816,
  "longitude" : -123.01101544961938,
  "altitude" : 1607.6589068715653,
  "altAgl" : 1607.6588833676299,
  "airspeed" : 52,
  "pitch" : 0.026875153183937073,
  "bank" : -2.423224722577144E-4,
  "heading" : 357.5731996999381,
  "aileron" : -6.103515625E-5,
  "elevator" : -6.103515625E-5,
  "rudder" : -6.103515625E-5,
  "throttle" : 73.236083984375
}
```
- See all the vors: [http://localhost:8080/vors](http://localhost:8080/vors)
```
[ {
  "icao" : "ITL",
  "latitude" : 49.19996581971645,
  "longitude" : -123.15604135394096,
  "altitude" : 4.267000198364258,
  "frequency" : 110550000,
  "localizer" : 100.01,
  "glideSlopeAngle" : 3.0,
  "glideLat" : 49.2051887512207,
  "glideLon" : -123.19614619016647,
  "glideAlt" : 4.267000198364258,
  "flags" : 15
}, {
...
```
- See all the airports: [http://localhost:8080/airports](http://localhost:8080/airports)
```
[ {
  "icao" : "WA77",
  "latitude" : 47.19565697014332,
  "longitude" : -122.02205613255501,
  "altitude" : 224.9420166015625
}, {
  "icao" : "7WA3",
  "latitude" : 48.88611115515232,
  "longitude" : -122.32861116528511,
  "altitude" : 48.768001556396484
}, {
  "icao" : "7WA5",
  "latitude" : 48.67287788540125,
  "longitude" : -123.1757453083992,
  "altitude" : 3.0480000972747803
}, {
...
```

Please note that data is limited to what FSX returns:
- aircrafts only exist in fsx inside a 199km radius around user aircraft
- same thing for airports and navigation aids, but since they have a fixed position,
they are stored forever in the cache since the flight start.

### Dependencies
FSX-saas depends on jSimconnect 0.8, which is **not** included with the code.
You can download it from the [original author site](http://lc0277.gratisim.fr/jsimconnect.html)
 or from [this github repository](https://github.com/mharj/jsimconnect).
 
It is needed only for development. Place it in the `lib/` folder and include it as a library in your IDE.
Library inclusion in the fat jar during build is managed by maven.

## Source code

If you want to develop on this codebase:
1. clone the repository into your IDE
2. download the jSimconnect jar and put it in the lib/ folder
 
## Feedback
If you think there is a bug in the program, or something is missing or wrong with the documentation/support files, feel free to [open an issue].

## License
This software is released under the LGPL V3 license.
The license terms are included in the LICENSE file.


[open an issue]: https://github.com/marcosox/fsx-saas/issues
[releases page]: https://github.com/marcosox/fsx-saas/releases


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [Unreleased]
### Added
- changelog file
### Changed
- nothing
### Deprecated
- nothing
### Removed
- nothing
### Fixed
- nothing
### Security
- nothing

## [1.2.0] - 2018-06-06
### Added
- METAR endpoint
- IFR flag for aircrafts and helicopters
### Changed
- nothing 
### Deprecated
- nothing
### Removed
- nothing
### Fixed
- incremented AI TRAFFIC STATUS string size
- copy items list instead of using reference when fetching it
### Security
- nothing

## [1.1.0] - 2018-04-30
### Added
- Helicopters
- Boats
- Ground vehicles
- Groundspeed info for aircrafts
### Changed
- Code refactoring
- Better values rounding
### Fixed
- Aircraft list size 0 did crash the app
- Command line examples in readme

## 1.0.0 - 2018-03-18
First version

[Unreleased]: https://github.com/marcosox/fsx-saas/compare/v1.2.0...HEAD
[1.2.0]: https://github.com/marcosox/fsx-saas/compare/v1.1.0...v1.2.0
[1.1.0]: https://github.com/marcosox/fsx-saas/compare/v1.0.0...v1.1.0

src/main/java/io/github/marcosox/fsxsaas/MainVerticle.java
package io.github.marcosox.fsxsaas;

import io.github.marcosox.fsxsaas.fsx.FSX;
import io.github.marcosox.fsxsaas.fsx.ObjectManager;
import io.vertx.core.AbstractVerticle;
import io.vertx.core.Future;
import io.vertx.core.Vertx;
import io.vertx.core.http.HttpHeaders;
import io.vertx.core.http.HttpMethod;
import io.vertx.core.json.Json;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.web.Route;
import io.vertx.ext.web.Router;
import io.vertx.ext.web.RoutingContext;
import io.vertx.ext.web.handler.BodyHandler;
import io.vertx.ext.web.handler.CorsHandler;

import java.io.IOException;
import java.util.List;
import java.util.Map;

public class MainVerticle extends AbstractVerticle {
	private final static String APP_NAME = "FSX simconnect as a service";
	private final static String APP_VERSION = "1.2.0";
	private final static int DEFAULT_PORT = 8080;
	private final static int DEFAULT_TRAFFIC_SCAN_INTERVAL_MS = 1000;

	private ObjectManager manager;
	private FSX fsx;

	/**
	 * Main entry point
	 *
	 * @param fut Vert.x Future object
	 */
	@Override
	public void start(Future<Void> fut) {
		System.out.println("Welcome to " + APP_NAME + " version " + APP_VERSION);
		System.out.println("MainVerticle.start()");
		setup();
		Router router = Router.router(vertx);
		router.route().handler(BodyHandler.create());
		router.route().handler(CorsHandler.create("*")
				.allowedMethod(HttpMethod.GET)
				.allowedMethod(HttpMethod.POST)
				.allowedHeader(HttpHeaders.ACCESS_CONTROL_ALLOW_ORIGIN.toString())
				.allowedHeader(HttpHeaders.CONTENT_TYPE.toString())
				.allowedHeader(HttpHeaders.ORIGIN.toString()));

		router.get("/aircrafts").handler(r -> this.handleGetAll(r, manager.getAircrafts()));
		router.get("/aircrafts/:id").handler(r -> this.handleGetItem(r, manager.getAircrafts()));
		router.get("/helicopters").handler(r -> this.handleGetAll(r, manager.getHelicopters()));
		router.get("/helicopters/:id").handler(r -> this.handleGetItem(r, manager.getHelicopters()));
		router.get("/boats").handler(r -> this.handleGetAll(r, manager.getBoats()));
		router.get("/boats/:id").handler(r -> this.handleGetItem(r, manager.getBoats()));
		router.get("/vehicles").handler(r -> this.handleGetAll(r, manager.getVehicles()));
		router.get("/vehicles/:id").handler(r -> this.handleGetItem(r, manager.getVehicles()));
		router.get("/airports").handler(r -> this.handleGetAll(r, manager.getAirports()));
		router.get("/airports/:id").handler(r -> this.handleGetItem(r, manager.getAirports()));
		router.get("/vors").handler(r -> this.handleGetAll(r, manager.getVors()));
		router.get("/vors/:id").handler(r -> this.handleGetItem(r, manager.getVors()));
		router.get("/ndbs").handler(r -> this.handleGetAll(r, manager.getNDBs()));
		router.get("/ndbs/:id").handler(r -> this.handleGetItem(r, manager.getNDBs()));
		router.get("/waypoints").handler(r -> this.handleGetAll(r, manager.getWaypoints()));
		router.get("/waypoints/:id").handler(r -> this.handleGetItem(r, manager.getWaypoints()));
		router.get("/metar").handler(this::handleMetar);
		router.get("/cmd/:command").handler(this::handleFsxCommand);
		router.get("/shutdown").handler(r -> this.quit(r, 0));
		List<Route> routes = router.getRoutes();
		router.get("/").handler(r -> this.handleRootURL(r, routes));

		int port = config().getInteger("port", DEFAULT_PORT);
		vertx.createHttpServer().requestHandler(router::accept).listen(port);
		System.out.println("HTTP server ready and listening on port " + port);
	}

	private void handleMetar(RoutingContext routingContext) {
		routingContext.response().putHeader("content-type", "text/plain").end(manager.getMetar());
	}

	/**
	 * Quits the application
	 *
	 * @param routingContext http request routing context
	 * @param status         exit status
	 */
	private void quit(RoutingContext routingContext, int status) {
		routingContext.response().putHeader("content-type", "text/plain").end("BYE");
		Vertx.vertx().close();
		System.exit(status);
	}

	/**
	 * Root url handler
	 *
	 * @param routingContext http request routing context
	 * @param routes         list of routes to display
	 */
	private void handleRootURL(RoutingContext routingContext, List<Route> routes) {
		StringBuilder routesList = new StringBuilder();
		routesList.append("<ul>");
		for (Route r : routes) {
			if (r.getPath() != null) {
				routesList.append("<li><a href='").append(r.getPath()).append("'>").append(r.getPath()).append("</a></li>");
			}
		}
		routesList.append("</ul>");
		routingContext.response().putHeader("content-type", "text/html").end(routesList.toString());
	}

	/**
	 * Handler for the LIST action
	 *
	 * @param routingContext http request routing context
	 * @param map            items map
	 */
	private void handleGetAll(RoutingContext routingContext, Map<String, ?> map) {
		routingContext.response().putHeader("content-type", "application/json").end(Json.encodePrettily(map.values()));
	}

	/**
	 * Handler for the RETRIEVE action
	 *
	 * @param r     http request routing context
	 * @param items items map
	 */
	private void handleGetItem(RoutingContext r, Map<String, ?> items) {
		String id = r.request().getParam("id");
		if (id != null) {
			Object item = items.get("" + id);
			if (item != null) {
				r.response().putHeader("content-type", "application/json").end(Json.encodePrettily(item));
			} else {
				if (id.equals("0")) {
					// user aircraft not found until sim is unpaused and the first info is sent
					r.response().setStatusCode(449).end("Unpause FSX for a second and try again.");
				} else {
					r.response().setStatusCode(404).end("item id " + id + " not found");
				}
			}
		}
	}

	/**
	 * Handles fsx commands: start/stop/restart/clear
	 *
	 * @param routingContext http request routing context
	 */
	private void handleFsxCommand(RoutingContext routingContext) {
		String cmd = routingContext.request().getParam("command");
		String msg;
		switch (cmd) {
			case "help": {
				msg = "<ul>" +
						"<li>start: start simconnect</li>" +
						"<li>stop: stop simconnect</li>" +
						"<li>restart: guess what</li>" +
						"<li>clear: clears all fsx objects from memory (aircrafts, vors, airports etc.)</li>" +
						"</ul>";
				break;
			}
			case "stop": {
				try {
					fsx.stopSimConnect();
					msg = "OK";
				} catch (IOException e) {
					e.printStackTrace();
					msg = e.getMessage();
				}
				break;
			}
			case "start": {
				try {
					fsx.startSimConnect();
					msg = "OK";
				} catch (IOException e) {
					e.printStackTrace();
					msg = e.getMessage();
				}
				break;
			}
			case "restart": {
				try {
					fsx.restartSimConnect();
					msg = "OK";
				} catch (IOException e) {
					e.printStackTrace();
					msg = e.getMessage();
				}
				break;
			}
			case "clear": {
				manager.clearAll();
				msg = "OK";
				break;
			}
			default: {
				msg = "not a command - try help";
				break;
			}
		}
		JsonObject responseMsg = new JsonObject();
		responseMsg.put("response", msg);
		routingContext.response().putHeader("content-type", "text/html").end(msg);
	}

	/**
	 * Starts simconnect and the object manager
	 */
	private void setup() {
		this.manager = new ObjectManager();
		this.fsx = new FSX(manager, config().getInteger("scanInterval", DEFAULT_TRAFFIC_SCAN_INTERVAL_MS));
	}
}
